### DISCUSSION TITLE
XQuery: ignore undeclared namespace prefix error

### DISCUSSION THREAD

**Discussion by fgeorges** [ajaxorg/ace#5255] [Q&A]
When using the XQuery mode, if a QName prefix has not been declared (e.g. in a function name), it is marked as an error:

![image](https://github.com/ajaxorg/ace/assets/584411/e0ad27d7-ba38-471b-ad33-b25fe582a1b9)

But some XQuery products have pre-defined namespaces (in XQuery parlance, they are declared implicitly in the static context.)  Is it possible to either:
- provide the XQuery mode with a list of pre-defined namespace prefixes?
- do not display the error for that specific case?

I tag @wcandillon and @nightwing, as you seem to be those who have worked on the XQuery mode, but it might be a general Ace question if the answer is not to display the error.

**Comment by fgeorges** [comment]
I forgot to post my sample repro:

```html
<!DOCTYPE html>
<html lang="en">
  <head>
    <title>Ace XQuery NS</title>
  </head>
  <body>
    <div id="editor">ns:f()</div>
    <script src="ace/ace.js" type="text/javascript" charset="utf-8"></script>
    <script>
      ace.edit('editor', {maxLines: 10, mode: 'ace/mode/xquery'});
    </script>
  </body>
</html>
```

**Comment by wcandillon** [comment]
hey Florent 🙋🏼‍♂️

This would need to be fixed on xqlint: https://github.com/wcandillon/xqlint
And then just use a version of the ace editor that ships this version of xqlint

**Comment by wcandillon** [comment]
This project might be the way to go: https://github.com/Quodatum/xqlint (cc @apb2006)

**Reply by apb2006** [reply]
Hi William, 
Thanks for the shoutout. 
I am working slowly on that Quodatum fork. My goals are BaseX support and targeting VSCode rather than Ace (https://github.com/Quodatum/vscode-basex).  I have removed the JSONiq support and more importantly here I have recently aligned the code "range" definition with that of VSCode.  So it is no longer a drop in replacement ( I somewhat regret this).

Hi Florent,
As William said it is the xqlint component that has the functionality to define the static context. I am not aware of a standard way to turn some or all error markers off at the Ace level.  It may be possible to monkey patch the code but in my experience it gets hard to reliably apply the patches before the rendering starts. If your goal is little more than your sample, ie a simple html page, it may be possible.

**Reply by fgeorges** [reply]
Hey @wcandillon 🙋🏼‍♂️!  Thank you for the pointer (not surprised you're behind this, to some extend :+1:)

Thank you, @apb2006!  Is it possible to access the `xqlint` objects when importing it through Ace? (maybe after importing explicitly `ace/mode/xquery`)  In order to add namespace prefix bindings to the static context before calling `ace.edit()`.

If adding namespaces to the static context used by `xqlint` is possible, and accessing the corresponding object is also possible, that would prevent me to have to maintain my own build.

If this is not possible but deemed a good addition, I am happy to try and help on this front, but I would need a couple of pointers in the code to get started, as an entry point.

**Reply by fgeorges** [reply]
So far, I could not find any other way to expand the list of predefined namespace prefixes, than modifying `ace/worker-xquery.js` and add extra such codes:

```js
namespaces['http://example.org/bar'] = {
    prefixes: ['bar'],
    pos: emptyPos,
    type: 'module',
    override: true
};
namespaces['http://example.org/foo'] = {
    prefixes: ['foo'],
    pos: emptyPos,
    type: 'module',
    override: true
};
```

Furthermore, because for some of the predefined prefixes, we sometimes need to import a module to access the functions it defines (you can use the prefix in an XPath expression for instance, but functions in that namespace still need to be imported,) I suppress the corresponding warning using the following:

```html
<!DOCTYPE html>
<html lang="en">
  <head>
    <title>Ace XQuery NS</title>
  </head>
  <body>
    <div id="editor">ns:f()</div>
    <script src="src/js/ace/ace.js" type="text/javascript" charset="utf-8"></script>
    <script>
      // the list of predefined prefixes, that also sometimes require to import a module
      const toignore = new Set();
      toignore.add('[W02] "http://example.org/bar" already bound to the "bar" prefix');
      toignore.add('[W02] "http://example.org/foo" already bound to the "foo" prefix');
      const ignore = msg => msg && toignore.has(msg);
      // load the xquery mode to override `addMarkers()`
      ace.config.loadModule('ace/mode/xquery', module => {
          // the XQuery mode constructor function
          const mode = module.Mode;
          // override `addMarkers()`, to ignore some specific warnings
          const orig = mode.prototype.addMarkers;
          mode.prototype.addMarkers = function(annos, session) {
              if ( ! (annos && ignore(annos[0]?.message)) ) {
                  return orig.call(this, annos, session);
              }
          };
      });
    </script>
    <script>
      ace.edit('editor', {maxLines: 10, mode: 'ace/mode/xquery'});
    </script>
  </body>
</html>
```

I don't think this is very clean to override `addMarkers()` like this, but at least I can do it from the outside of Ace/XQLint code.

I could not find any such way for adding namespace prefixes to the static context, without modifying `ace/worker-xquery.js` directly.  Do you think there is a way?

**Reply by apb2006** [reply]
Hi @fgeorges, 
I think what you show is probably about the best that can be done without code changes.
There is a call on the StaticContext to add a namespace
https://github.com/wcandillon/xqlint/blob/master/lib/compiler/static_context.js#L198
but I can not see a way to call that from the ace level.
The `worker-xquery.js` has
```
var XQueryWorker = exports.XQueryWorker = function(sender) {
    Mirror.call(this, sender);
    this.setTimeout(200);
    var that = this;

    this.sender.on("complete", function(e){
        if(that.xqlint) {
            var pos = { line: e.data.pos.row, col: e.data.pos.column };
            var proposals = that.xqlint.getCompletions(pos);
            that.sender.emit("complete", proposals);
        }
    });

    this.sender.on("setAvailableModuleNamespaces", function(e){
        that.availableModuleNamespaces = e.data;
    });

    this.sender.on("setFileName", function(e){
        that.fileName = e.data;
    });

    this.sender.on("setModuleResolver", function(e){
        that.moduleResolver = getModuleResolverFromModules(e.data);
    });
};
```
But I think `availableModuleNamespaces ` is just used for completion. `namespaces` is what you need for validation. 
`setModuleResolver` could be used to add functions but this only works properly if the namespace is defined first.

Note also the XQuery parser used here is for 3.0 not 3.1 so no `map{}`s etc.
/Andy
